L1Trigger/TrackFindingTracklet/src/VMRouterCM.cc

```diff
@@ -0,0 +1,283 @@
+      if (seedindex == -1) {
+        seedindex = vmstubsTEPHI_.size();
+        vector<VMStubsTEMemory*> vectmp;
+        VMStubsTEPHICM atmp(iseed, vectmp);
+        vmstubsTEPHI_.push_back(atmp);
+      }
+      tmp->resize(settings_.NLONGVMBINS() * settings_.nvmte(1, iseed));
+      vmstubsTEPHI_[seedindex].vmstubmem.push_back(tmp);
+
+    } else if (memory->getName().substr(3, 2) == "ME") {
+      VMStubsMEMemory* tmp = dynamic_cast<VMStubsMEMemory*>(memory);
+      assert(tmp != nullptr);
+      tmp->resize(nvmmebins_ * settings_.nvmme(layerdisk_));
+      assert(vmstubsMEPHI_[0] == nullptr);
+      vmstubsMEPHI_[0] = tmp;
+    } else {
+      throw cms::Exception("LogicError") << __FILE__ << " " << __LINE__ << " should never get here!";
+    }
+
+    return;
+  }
+
+  throw cms::Exception("BadConfig") << __FILE__ << " " << __LINE__ << " Could not find output : " << output;
+}
+
+void VMRouterCM::addInput(MemoryBase* memory, string input) {
+  if (settings_.writetrace()) {
+    edm::LogVerbatim("Tracklet") << "In " << name_ << " adding input from " << memory->getName() << " to input "
+                                 << input;
+  }
+  if (input == "stubin") {
+    InputLinkMemory* tmp1 = dynamic_cast<InputLinkMemory*>(memory);
+    assert(tmp1 != nullptr);
+    if (tmp1 != nullptr) {
+      stubinputs_.push_back(tmp1);
+    }
+    return;
+  }
+  throw cms::Exception("BadConfig") << __FILE__ << " " << __LINE__ << " Could not find input : " << input;
+}
+
+void VMRouterCM::execute() {
+  unsigned int allStubCounter = 0;
+
+  //Loop over the input stubs
+  for (auto& stubinput : stubinputs_) {
+    for (unsigned int i = 0; i < stubinput->nStubs(); i++) {
+      if (allStubCounter > settings_.maxStep("VMR"))
+        continue;
+      if (allStubCounter > 127)
+        continue;
+
+      Stub* stub = stubinput->getStub(i);
+
+      //Note - below information is not part of the stub, but rather from which input memory
+      //we are reading
+      bool negdisk = (stub->disk().value() < 0);
+
+      //use &127 to make sure we fit into the number of bits -
+      //though we should have protected against overflows above
+      FPGAWord allStubIndex(allStubCounter & 127, 7, true, __LINE__, __FILE__);
+
+      //TODO - should not be needed - but need to migrate some other pieces of code before removing
+      stub->setAllStubIndex(allStubCounter);
+      //TODO - should not be needed - but need to migrate some other pieces of code before removing
+      stub->l1tstub()->setAllStubIndex(allStubCounter);
+
+      allStubCounter++;
+
+      FPGAWord iphi = stub->phicorr();
+      unsigned int iphipos = iphi.bits(iphi.nbits() - (settings_.nbitsallstubs(layerdisk_) + 3), 3);
+
+      //Fill allstubs memories - in HLS this is the same write to multiple memories
+      for (auto& allstub : allstubs_) {
+        char memtype = allstub.first;
+        if (memtype == 'R' && iphipos < 5)
+          continue;
+        if (memtype == 'L' && iphipos >= 3)
```

Comment:
these could all be combined into a single if statement

Reply:
fixed